libp2p/autotls/acme/api.nim

```diff
@@ -0,0 +1,387 @@
+import options, base64, sequtils, json
+from times import DateTime, parse
+import chronos/apps/http/httpclient, jwt, results, bearssl/pem
+
+import ./utils
+import ../../crypto/crypto
+import ../../crypto/rsa
+import ../../transports/tls/certificate_ffi
+import ../../transports/tls/certificate
+
+const
+  LetsEncryptURL* = "https://acme-v02.api.letsencrypt.org"
+  LetsEncryptURLStaging* = "https://acme-staging-v02.api.letsencrypt.org"
+  Alg = "RS256"
+  DefaultChalCompletedRetries = 10
+  DefaultChalCompletedRetryTime = 1.seconds
+  DefaultFinalizeRetries = 10
+  DefaultFinalizeRetryTime = 1.seconds
+  DefaultRandStringSize = 256
+
+type Nonce = string
+type Kid = string
+type AccountStatus = string
+
+type ACMEDirectory = object
+  newNonce: string
+  newOrder: string
+  newAccount: string
+
+type ACMEApi* = object
+  directory: ACMEDirectory
+  session*: HttpSessionRef
+  acmeServerURL: string
+
+type ACMEResponse = object
+  body: JsonNode
+  headers: HttpTable
+
+type JWK = object
+  kty: string
+  n: string
+  e: string
+
+# whether the request uses Kid or not
+type ACMERequestType = enum
+  ACMEJwkRequest
+  ACMEKidRequest
+
+type ACMERequestHeader = object
+  alg: string
+  typ: string
+  nonce: string
+  url: string
+  case kind: ACMERequestType
+  of ACMEJwkRequest:
+    jwk: JWK
+  of ACMEKidRequest:
+    kid: Kid
+
+type ACMERegisterRequest* = object
+  termsOfServiceAgreed: bool
+  contact: seq[string]
+
+type ACMERegisterResponse* = object
+  kid*: Kid
+  status*: AccountStatus
+
+type ACMEChallenge = object
+  url*: string
+  `type`*: string
+  status*: string
+  token*: string
+
+type ACMEChallengeIdentifier = object
+  `type`: string
+  value: string
+
+type ACMEChallengeRequest = object
+  identifiers: seq[ACMEChallengeIdentifier]
+
+type ACMEChallengeResponse* = object
+  status: string
+  authorizations: seq[string]
+  finalize: string
+
+type ACMEChallengeResponseWrapper* = object
+  finalizeURL*: string
+  orderURL*: string
+  dns01*: ACMEChallenge
+
+type ACMEAuthorizationsResponse* = object
+  challenges: seq[ACMEChallenge]
+
+type ACMECompletedResponse* = object
+  checkURL: string
+
+type ACMECheckResponse* = object
+  status: string
+
+type ACMEFinalizedResponse* = object
+
+type ACMEOrderResponse* = object
+  certificate: string
+  expires: string
+
+type ACMECertificateResponse* = object
+  rawCertificate: string
+  certificateExpiry: DateTime
+
+template handleError(msg: string, body: untyped): untyped =
```

Comment:
`handleError` must propagate CancelledError, otherwise it will be caught by `CatchableError`